### ISSUE TITLE
[Checklist] More signatures to-do items

### ISSUE THREAD

**Issue by Maikuolan** [phpMussel/phpMussel#124]
- [x] Devise a better method for cutting new ClamAV signature sets for phpMussel (developing new automation tools the preferred method; need to plan this properly; high priority).
- [x] Begin writing new automation tools and/or other necessary tools for the new method.
- [x] Test and confirm operability of new tools/method.
- [x] Since migrating the signature files into their repository, no new cuts of the ClamAV signature sets have been produced. This is currently pending on devising a better method for doing so. Need to get a new cut out ASAP after having devised aforementioned better method / writing aforementioned tools.
- [x] Update README documentation to reflect all above-mentioned changes.

**Comment by wajdijurry** [comment]
Hi Caleb, 
So when do you expect the release date ?

**Comment by Maikuolan** [comment]
> Hi Caleb,
> So when do you expect the release date ?

Truthfully, I don't know. I hope sooner, rather than later (I'd be happy to release it today, right now, if it was ready for that); Especially seeing as now more than a year has passed since the previous release (February 27, 2016), but at the same time, I'd prefer to do it properly if at all possible, rather than to release something half-arsed or broken; If `v1.0.0` is released and users immediately notice major problems, or something inherently broken with the way that signatures are cut and included in the package, or if the new package release simply fails to do what it claims to be able to do, I think that those things would be more damaging to the package than simply releasing it a few months later than initially anticipated.

The first item in the above list (*"Devise a better method for cutting new ClamAV signature sets for phpMussel"*) is the most important of the lot, and involves more planning and consideration than actual coding or anything else, per se. Foremost, in that all ClamAV signatures, collectively, total nearly half a GB at this point, the best way to process them all in a way that won't just crash the PHP process entirely, that can be done in a timely manner, and that will actually work properly for our purposes; Not only processing them, but the process of deciding what to keep and what to ditch - For most shared hosting solutions and other similarly positioned websites, the owners of said websites won't be wanting a nearly half a GB added to the overall footprint of their website simply for the sake of being able to scan file uploads, which means that when figuring out the best method for automating the process for cutting signature files from the ClamAV database for use with phpMussel, consideration will need to be made for being able to automatically weigh up the value of different signatures within the ClamAV database to determine whether or not they should be included with the phpMussel signatures. This is also partly why I'd abandoned previous plans to implement support for additional types of signatures in the future (like icon and certificate scanning, for example); That would likely only add to a footprint which would already need to be slimmed down quite a bit. These considerations also bring into question whether the actual format for phpMussel signatures should just be changed, to slim it down and increase efficiency, which in-turn, would mean further rewrites of the core code anyway, and possible additional backwards-incompatible changes, which would be preferable to avoid if at all possible, but are nonetheless things to think about, in any case.

A more intelligent approach to heuristic analysis and a better ability to analyse file containers of various formats would probably significantly cut down the number of signatures needed to be included in order for the package to be vaguely useful, and is definitely something desirable in any case, but of course, prioritising this would mean further pushing back any plans to devise a better method for cutting new signatures, which would mean further pushing back the eventual `v1.0.0` release.

Don't get the wrong idea, of course; I'm not complaining about any of this, I still intend to eventually get it all done, and I have no intentions of complaining about things being difficult, but at the same time, comparing phpMussel against some of the other open-source projects I'm managing at the moment (like CIDRAM, for example), the problems facing phpMussel are definitely more complex.

This is also one of the reasons I kind of like lists like these: They help to keep a bit of perspective, and prevent me from forgetting things (usually, anyhow).

Anyhow. Sorry for not giving a clear, exact date for this, and I hope the above explanation goes some way to explaining why I haven't given a clear, exact date for this. An optimistic date would be the end of this month (a date that, if I have plenty of free time, plenty of good luck, manage to figure everything out, and if nothing at all goes wrong in that time, I *might* be able to reach). A more realistic date (assuming delays, things going wrong, and a general lack of free time) would be two months or so. If it goes beyond that and starts edging towards the end of the year, I might need to start considering handing the project off to someone else that could do a better job of it than I've been able to and then just let it go and move on. After all.. Despite the time and effort that I've put into the project thus far, and with no disrespect intended towards it, it isn't something that I necessarily want to be tied down to for the rest of my life or for the rest of my time as a programmer; I do hope to eventually call it finished and move on, one way or another. But, we're not quite there yet, I think. ;-)

I guess, depending on the types of files you were needing to scan, I could probably do a quick once off interim cut for whichever signature files were needed, using the latest ClamAV database; I'd rather not do the whole lot, because cutting the whole lot manually takes quite a bit of time, and would hopefully, eventually, become pointless, once an automated method has been put in place, but if it's just for a small subsection of signatures for some specific types of files or something like that, I might be able to write some up quickly.

**Comment by Maikuolan** [comment]
*Addendum: Direct assistance from others where possible, may help to cut down the required time quite a bit, and result in an earlier release date.*

**Comment by Maikuolan** [comment]
I've started coding up a tool in C# that I'm hoping should be able to achieve what we're wanting here. Nothing ready to test yet (only just starting), but I'll post here again when I get a bit further with it.

*Removed Planning label and added Work-In-Progress label.*

**Comment by DanielRuf** [comment]
Personally I would prefer Java as it runs on all systems but great to hear that we are getting a tool (should be easy to migrate the code to Java).

**Comment by nhlm** [comment]
Why not using PHP for that?

**Comment by DanielRuf** [comment]
Or PHP for the CLI (as Phar / executable bin/php). Java was just a suggestion instead of C# as both have a very similar API ;-)

**Comment by Maikuolan** [comment]
My main reason against using PHP is the problems inherent with attempting to unzip nearly ~500MBs worth of raw data (most of which is ultimately ditched during the process anyhow) using the in-built PHP ZIP functions (exceeding memory limitations, randomly crashing, etc; hasn't always reliable whenever I've tested it).

That said though, I've ran into a minor problem while trying to build a tool in C# to handle this (definitely not insurmountable nor difficult to work around, but kind of somewhat tedious); Aside from being locked to Windows systems, using ZIP functions in C# would also lock any said tool to systems with at least version 4.5 of the .NET Framework or higher installed (v4.5 is the lowest version to support basic ZIP functions), which is a little newer than I was hoping to have as a minimum required version if I was to be using C# to build this (I have a Windows 10 machine with the latest .NET installed, but the minimum required version isn't supported by the Windows 7 machine that I prefer to use for all my coding whenever possible, for whatever reason; it *should* be compatible according to the .NET documentation, but says otherwise when I attempt to install it). Still doable, but the inability to execute said tool on anything but the very latest Windows machines isn't particularly appealing, and I'd absolutely like to avoid needing to port/polyfill basic ZIP functions into the tool's code itself if at all possible, regardless of what language is ultimately used (C#, PHP, Java, whatever); Not reinventing the wheel and all that.

So, for now, I've put the C# plan on hold, and I'm back to thinking on this again, as to what would be the best way to go about this.

Investigating Java as an alternate possibility, and otherwise, still investigating what can be done about the problems encountered when trying to do this using PHP; If it can't be done in a clean, straightforward way, a last ditch resort might be to just tie it in with an external binary to handle unzipping the archives, external binaries to handle whatever other problematic parts of the process might appear, and just use PHP basically the stitch it all together and supervise the whole process (which might ultimately add some configuration burden to the tool, but would probably mean less coding to do in the long run anyhow, I guess).

**Comment by DanielRuf** [comment]
Hm, batch / shell script which executes 7zip.exe (standalone) and does the other tasks? Java so far has less dependencies, there are great libraries and can consume much RAM if needed.

**Comment by DanielRuf** [comment]
Or write some small Node.js script, should run on all platforms, does not have to be compiled and has a huge amount of modules (see npmjs.com) and is fast. Or Python? Not sure but personally I would prefer Node.js for new tools.

**Comment by Maikuolan** [comment]
Well.. Not sure whether I was just doing it a really, really dumb way before, whether I can attribute it to improvements with PHP7, or whatever else, but I seem to be having a lot more luck this time around than before, just doing it in plain PHP. Although neither thoroughly tested nor documented yet, I've got something *kind* of workable, which seems okay thus far, and is able to produce signature files for phpMussel from the ClamAV signature files without any real difficulty, and in basically under a minute, which is.. quite a bit better than the good hour or so that my previous attempt from last year required, when it didn't just outright crash. No crashing at all so far with the new tool. New tool is still alpha and is ugly as sin, but I'll make it prettier after it's been tested a bit and is properly working as intended (for now, I'm just wanting to get it working as intended, and then it can be improved on from there).

Syncing a commit with a new cut of signatures right now (untested!), which should be finished syncing very soon, hopefully.

**Comment by wajdijurry** [comment]
Hi Clab,

Any updates about phpMussel viruses database update ? I am waiting the
release to take the latest update of phpMussel antivirus.

Thanks,

**Comment by Maikuolan** [comment]
The latest cut that I released just now *looks* okay, but I haven't tested it, so it's possible (note: quite likely) that unexpected problems may (will) emerge. I plan to test it tomorrow, and if there aren't any unexpected problems, I'll start working on documenting it and setting up appropriate deployment channels to make it easier to update to the latest cuts, and otherwise, if there are unexpected problems, I'll keep working at it to try to resolve them.

In theory, the latest cut could *probably* be used immediately, but I don't really like green-lighting something as being okay without actually testing it first, and so, as a conservative estimate, I'd say a couple of days. ;-)

**Comment by wajdijurry** [comment]
Is the latest release version of phpMussel is 0.10.0 or there is another
version ? If there please provide me with.

Thanks,

**Comment by Maikuolan** [comment]
Officially, yes, 0.10.0 is the latest release version. The latest cut is designed to work with 1.0.0-DEV though (the current latest codebase available in the repository master branch), and won't work with earlier (or the official latest) version. 1.0.0-DEV (or 1.0.0, when actually officially released), is available (that's what'll be downloaded when you download the latest codebase directly as a ZIP or TAR file), but hasn't yet been *officially* released. It'll be officially released once everything has been tested (including latest signatures), confirmed to be working as intended, once latest changes have finished being documented and a few pending bugs and already planned small improvements have been sorted out (so, soon).

At this point, if we think of any completely new features that haven't already been planned or started, they'll probably get pushed to v1.1.0 or something I reckon.

**Comment by Maikuolan** [comment]
Well.. A very brief test before I get some sleep for the night:

![test](https://user-images.githubusercontent.com/12571108/29041737-9b921f8a-7be6-11e7-9524-d38af02e9a1d.png)

New cut seems to work correctly, but it's disgustingly slow. I'll work on some optimisations to the signature tool tomorrow to try to generate a cut that works a bit faster.

The standard testfiles included with the package seem to scan successfully within less than a few seconds, as do most files under the ~100KB mark; From there, it progressively gets slower and slower, the larger the file.

For the two files scanned as per the attached screenshot, the first is a 1,350 KB PE file (the 7-Zip installer) and the second is a 34,604 KB JPEG file.

Scanned a few hundred various other files too, all with very predictably similar results. Summary: Technically, the new cut is operable, but optimisations would definitely be a good idea.

Throughout the whole process, no crashing occurred (so, it seems to be crash-safe, at least). However, testing was all done via CLI-mode, with no maximum execution time or timeouts applied (scanning via a web interface or via a normal file upload may yield very different results, and considering that the latter of the two examples included in the screenshot required slightly over 3 minutes to finish scanning, it's more than likely that a timeout would occur if it was scanned as a normal file upload).

Will continue tomorrow. Goodnight.

*Note: Above testing was done with ALL signature files as ACTIVE, and with all blacklists, whitelists, and greylists as DISABLED. Required scan time could be improved for specific installations by activating only the signature files required as per the installation setup, accounting for blacklists, whitelists and so on, and with unrequired filetypes being blacklisted and such.*

**Comment by Maikuolan** [comment]
Probably will make continual improvements over time, but basically, it works. Will get started on documentation later today.

**Comment by Maikuolan** [comment]
*Documentation updated. All items implemented. Marking and closing.*
